File: leetcode/276. Paint Fence/276.cpp
// https://helloacm.com/compute-the-number-of-ways-to-paint-the-house-via-dynamic-programming-algorithm/
// https://leetcode.com/problems/paint-fence/

class Solution {
public:
    int numWays(int n, int k) {
        vector<int> f(n + 1);
        f[0] = 0;
        if (n == 0) return f[0];
        f[1] = k;
        if (n == 1) return f[1];
        f[2] = k * k;
        if (n == 2) return f[2];
        for (int i = 3; i <= n; ++ i) {
            f[i] = f[i - 1] * (k - 1) + // different color 
                f[i - 2] * (k - 1); // same color
        }
        return f[n];
    }
};

File: leetcode/276. Paint Fence/276.py
// https://helloacm.com/teaching-kids-programming-paint-fences-no-3-consecutive-same-colours-top-down-dynamic-programming-algorithm-recursion-memoization/
// https://leetcode.com/problems/paint-fence/

class Solution:
    def numWays(self, n: int, k: int) -> int:
        if n == 0:
            return 0
        if n == 1:
            return k        
        a, b = k, k * k
        for i in range(3, n + 1):
            a, b = b, (k - 1) * (a + b)
        return b
